src/psyclone/psyir/transformations/replace_induction_variables.py

```diff
@@ -0,0 +1,272 @@
+
+        :param assignment: the assignment statement to be tested, which \
+            must not be an array variable.
+        :type assignment: :py:class:`psyclone.psyir.nodes.Assignment`
+        :param accesses_in_loop_body: the access information for all \
+            variables in the loop body.
+        :type accesses_in_loop_body: \
+            :py:class:`psyclone.core.VariablesAccessInfo`
+
+        :returns: whether the assignment is an induction statement that \
+            can be replaced.
+        :rtype: bool
+
+        '''
+        # Collect all variables used on the rhs of assignment:
+        rhs_accesses = VariablesAccessInfo(assignment.rhs)
+        # If the rhs uses any variable that is written in the loop body, this
+        # is not a (simple) induction variable and cannot be replaced.
+        # Note that the write to the loop variable is part of the Loop
+        # statement, and is therefore not included in accesses_in_loop_body.
+        # This statement will also catch assignments like 'k=k+3' (which
+        # could be support in some cases):
+        if any(not accesses_in_loop_body[sig].is_read_only()
+               for sig in rhs_accesses):
+            return False
+
+        # Test that this assignment is the first access to this variable
+        # in the body. Get the signature from the node (ignore indices)
+        sig = assignment.lhs.get_signature_and_indices()[0]
+        var_accesses = accesses_in_loop_body[sig]
+        if var_accesses[0].node is not assignment.lhs:
+            return False
+
+        # Check if there is another write to this variable
+        # after the first one:
+        if any(access.access_type != AccessType.READ
+               for access in var_accesses.all_accesses[1:]):
+            return False
+
+        return True
+
+    # ------------------------------------------------------------------------
+    def apply(self, node, options=None):
+        '''Apply the ReplaceInductionVariables transformation to the
+        specified node. The node must be a loop. In case of nested
+        loops, the transformation might need to be applied several
+        time, from the inner-most loop outwards.
+
+        :param node: a Loop node.
+        :type node: :py:class:`psyclone.psyir.nodes.Loop`
+
+        '''
+        self.validate(node)
+        loop_var = node.variable.name
+
+        # Find assignments that are directly part of the loop (this avoid
+        # issues with assignment inside if statements):
+        all_accesses = VariablesAccessInfo(node.loop_body)
+        indx = 0
+        while indx < len(node.loop_body.children):
+            assignment = node.loop_body.children[indx]
+            # Only handle assignments, ignore if statements etc
+            if not isinstance(assignment, Assignment):
+                indx += 1
+                continue
+
+            # Assignment to arrays are ignored as well:
+            if isinstance(assignment.lhs, ArrayReference):
```

Comment:
This won't catch Structure of arrays, e.g. "var.field(3) = 4" is this a problem/. In the transformation docstring you mention that this applies to "non-structure scalars" but I can't see where this is checked, shouldn't this be one of this checks with "indx+=1;continue"?

Reply:
No idea what happened here - I was certain I had a a test for arrays, and somehow thought I tested for References (not instance of Reference).
I've fixed it all - it now supports structures, as long as there is no index used anywhere. Tests also added.
Note that I could not get the parser to handle a function call. So in the corresponding test for function calls on the RHS I had to manually modify the PSyIR to insert a 'Call' (which is a function call). Do we need a ticket for the issue with function calls?

Reply:
This is apparently a known issue, so this is ok.